### PR TITLE
[-Wunsafe-buffer-usage] Fix a bug and suppress libc warnings for C files

### PR DESCRIPTION
**Pull request by ziqingluo-90**
- Fix a bug in UnsafeBufferUsage.cpp related to casting to PointerType (report by @hnrklssn [here](https://github.com/llvm/llvm-project/pull/101583#discussion_r1767705992))
- Suppress -Wunsafe-buffer-usage-in-libc-call for C files

### CODE REVIEW

File: clang/lib/Analysis/UnsafeBufferUsage.cpp [hunk 1]
```diff
@@ -789,7 +791,7 @@ AST_MATCHER_P(CallExpr, hasUnsafePrintfStringArg,
   if (!FristParmTy->isPointerType())
     return false; // possibly some user-defined printf function
 
-  QualType FirstPteTy = (cast<PointerType>(FristParmTy))->getPointeeType();
+  QualType FirstPteTy = FristParmTy->getAs<PointerType>()->getPointeeType();
```
Comment (hnrklssn) [inline clang/lib/Analysis/UnsafeBufferUsage.cpp:794]: Nit: FristParmTy -> FirstParmTy

**Review by hnrklssn — COMMENTED** [review]
You may want to add some tests with annotated pointers to verify that you treat them correctly, e.g. `_Nullable`

File: clang/test/SemaCXX/warn-unsafe-buffer-usage-no-libc-functions-in-c.c [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang_cc1 -Wunsafe-buffer-usage %s -verify %s
```
Comment (jkorous-apple) [inline clang/test/SemaCXX/warn-unsafe-buffer-usage-no-libc-functions-in-c.c:1]: FWIW the language mode is set via `-x` command line option. It is only inferred from the file extension if it's missing.

File: clang/lib/Analysis/UnsafeBufferUsage.cpp [hunk 3]
```diff
@@ -250,7 +250,9 @@ AST_MATCHER_P(Stmt, ignoreUnsafeBufferInContainer,
 
 AST_MATCHER_P(Stmt, ignoreUnsafeLibcCall, const UnsafeBufferUsageHandler *,
               Handler) {
-  return Handler->ignoreUnsafeBufferInLibcCall(Node.getBeginLoc());
+  if (Finder->getASTContext().getLangOpts().CPlusPlus)
+    return Handler->ignoreUnsafeBufferInLibcCall(Node.getBeginLoc());
+  return true; /* Only warn about libc calls for C++ */
```
Comment (jkorous-apple) [inline clang/lib/Analysis/UnsafeBufferUsage.cpp:255]: This implementation should warn in Obj-C++ too and it should not in Obj-C but you might want to test that.

File: clang/lib/Sema/AnalysisBasedWarnings.cpp [hunk 4]
```diff
@@ -2549,6 +2549,7 @@ void clang::sema::AnalysisBasedWarnings::IssueWarnings(
   DiagnosticOptions &DiagOpts = Diags.getDiagnosticOptions();
 
   // UnsafeBufferUsage analysis settings.
+  bool IsCXXLang = S.getLangOpts().CPlusPlus;
```
Comment (jkorous-apple) [inline clang/lib/Sema/AnalysisBasedWarnings.cpp:2552]: Nit: it doesn't feel like we need a variable for that.

**Review by jkorous-apple — APPROVED** [review]
Left some comments but LGTM.

File: clang/lib/Analysis/UnsafeBufferUsage.cpp [hunk 5]
```diff
@@ -784,12 +786,12 @@ AST_MATCHER_P(CallExpr, hasUnsafePrintfStringArg,
...
   ASTContext &Ctx = Finder->getASTContext();
-  QualType FristParmTy = FD->getParamDecl(0)->getType();
+  QualType FirstParmTy = FD->getParamDecl(0)->getType();
 
-  if (!FristParmTy->isPointerType())
+  if (!FirstParmTy->isPointerType())
     return false; // possibly some user-defined printf function
 
-  QualType FirstPteTy = (cast<PointerType>(FristParmTy))->getPointeeType();
+  QualType FirstPteTy = FirstParmTy->getAs<PointerType>()->getPointeeType();
```
Comment (haoNoQ) [inline clang/lib/Analysis/UnsafeBufferUsage.cpp:794]: `castAs`?

**Review by haoNoQ — APPROVED** [review]
LGTM!

File: clang/lib/Analysis/UnsafeBufferUsage.cpp (hunk 5 above)
Comment (ziqingluo-90) [reply clang/lib/Analysis/UnsafeBufferUsage.cpp]: The difference between `castAs` and `getAs` is that `castAs` asserts the canonical type is `PointerType` while `getAs` returns null if not.  I think `castAs` is more suitable for our case.

**Pull request merged**